File: hermes/plugin/_tools_input.py
"""Keyboard and mouse input tools for the deskbrid Hermes plugin."""
from __future__ import annotations

import json

from ._socket import send_action


def _register_input_tools(ctx):
    """Keyboard and mouse input tools."""
    ctx.register_tool(
        name="type_text",
        toolset="deskbrid",
        schema={
            "name": "type_text",
            "description": "Type a string of text via keyboard input. Types at human speed.",
            "parameters": {
                "type": "object",
                "properties": {"text": {"type": "string", "description": "Text to type"}},
                "required": ["text"],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({"type": "input.keyboard", "id": "1", "action": "type", "text": params.get("text", "")})
        ),
        description="Type text via keyboard input.",
    )

    ctx.register_tool(
        name="press_key",
        toolset="deskbrid",
        schema={
            "name": "press_key",
            "description": "Press and release a single key by name (e.g., 'Return', 'Escape', 'Tab', 'F5').",
            "parameters": {
                "type": "object",
                "properties": {"key": {"type": "string", "description": "Key name (X11 keysym)"}},
                "required": ["key"],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({"type": "input.keyboard", "id": "1", "action": "key", "key": params.get("key", "")})
        ),
        description="Press a single key.",
    )

    ctx.register_tool(
        name="press_keys",
        toolset="deskbrid",
        schema={
            "name": "press_keys",
            "description": "Press a key combination. Keys pressed in order, released in reverse.",
            "parameters": {
                "type": "object",
                "properties": {
                    "keys": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": "Keys to press, e.g., ['Control_L', 'c']",
                    },
                },
                "required": ["keys"],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({
                "type": "input.keyboard", "id": "1", "action": "combo",
                "keys": params.get("keys", []),
            })
        ),
        description="Press a key combination.",
    )

    ctx.register_tool(
        name="mouse_move",
        toolset="deskbrid",
        schema={
            "name": "mouse_move",
            "description": "Move the mouse cursor to absolute screen coordinates.",
            "parameters": {
                "type": "object",
                "properties": {
                    "x": {"type": "number", "description": "X coordinate (pixels)"},
                    "y": {"type": "number", "description": "Y coordinate (pixels)"},
                },
                "required": ["x", "y"],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({"type": "input.mouse", "id": "1", "action": "move", "x": params.get("x", 0), "y": params.get("y", 0)})
        ),
        description="Move the mouse cursor.",
    )

    ctx.register_tool(
        name="mouse_click",
        toolset="deskbrid",
        schema={
            "name": "mouse_click",
            "description": "Click a mouse button at the current cursor position.",
            "parameters": {
                "type": "object",
                "properties": {
                    "button": {
                        "type": "string",
                        "enum": ["left", "middle", "right"],
                        "description": "Mouse button to click",
                    },
                },
                "required": ["button"],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({"type": "input.mouse", "id": "1", "action": "click", "button": params.get("button", "left")})
        ),
        description="Click a mouse button.",
    )

    ctx.register_tool(
        name="mouse_scroll",
        toolset="deskbrid",
        schema={
            "name": "mouse_scroll",
            "description": "Scroll the mouse wheel. Positive dy = scroll down, negative = scroll up.",
            "parameters": {
                "type": "object",
                "properties": {
                    "dx": {"type": "number", "description": "Horizontal scroll (default: 0)"},
                    "dy": {"type": "number", "description": "Vertical scroll (positive = down)"},
                },
                "required": [],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({"type": "input.mouse", "id": "1", "action": "scroll", "dx": params.get("dx", 0), "dy": params.get("dy", 0)})
        ),
        description="Scroll the mouse wheel.",
    )

    ctx.register_tool(
        name="click_coordinate",
        toolset="deskbrid",
        schema={
            "name": "click_coordinate",
            "description": "Move to pixel coordinates and click. Combines mouse_move + mouse_click.",
            "parameters": {
                "type": "object",
                "properties": {
                    "x": {"type": "number"},
                    "y": {"type": "number"},
                    "button": {"type": "string", "enum": ["left", "middle", "right"]},
                },
                "required": ["x", "y"],
            },
        },
        handler=lambda params, **kw: json.dumps(
            send_action({
                "type": "input.mouse", "id": "1", "action": "click",
                "x": params.get("x", 0), "y": params.get("y", 0),
                "button": params.get("button", "left"),
            })
        ),
        description="Click at specific screen coordinates.",
    )

File: src/mcp/tools_input.rs
#[macro_export]
macro_rules! tools_input {
    () => {

    #[tool(
        name = "type_text",
        description = "Type a string via keyboard input.",
        annotations(
            read_only_hint = false,
            destructive_hint = true,
            idempotent_hint = false,
            open_world_hint = true
        )
    )]
    async fn type_text(&self, Parameters(TypeText { text }): Parameters<TypeText>) -> String {
        self.call(do_type_text(&self.state, &text)).await
    }

    #[tool(
        name = "press_key",
        description = "Press a single key (e.g. 'Return', 'Escape', 'Tab').",
        annotations(
            read_only_hint = false,
            destructive_hint = true,
            idempotent_hint = false,
            open_world_hint = true
        )
    )]
    async fn press_key(&self, Parameters(PressKey { key }): Parameters<PressKey>) -> String {
        self.exec("input.keyboard", json!({"key": key}),).await
    }

    #[tool(
        name = "press_keys",
        description = "Press a key combination (e.g. ['Control_L', 'c'] for Ctrl+C).",
        annotations(
            read_only_hint = false,
            destructive_hint = true,
            idempotent_hint = false,
            open_world_hint = true
        )
    )]
    async fn press_keys(&self, Parameters(PressKeys { keys }): Parameters<PressKeys>) -> String {
        self.call(do_press_keys(&self.state, &keys)).await
    }

    #[tool(
        name = "mouse_move",
        description = "Move the mouse cursor to absolute coordinates.",
        annotations(
            read_only_hint = false,
            destructive_hint = false,
            idempotent_hint = true,
            open_world_hint = true
        )
    )]
    async fn mouse_move(&self, Parameters(MouseMove { x, y }): Parameters<MouseMove>) -> String {
        self.call(do_mouse_move(&self.state, x, y)).await
    }

    #[tool(
        name = "mouse_click",
        description = "Click a mouse button at the current position.",
        annotations(
            read_only_hint = false,
            destructive_hint = true,
            idempotent_hint = false,
            open_world_hint = true
        )
    )]
    async fn mouse_click(
        &self,
        Parameters(MouseClick { button }): Parameters<MouseClick>,
    ) -> String {
        self.call(do_mouse_click(&self.state, &button)).await
    }

    #[tool(
        name = "mouse_scroll",
        description = "Scroll the mouse wheel. Negative dy scrolls down.",
        annotations(
            read_only_hint = false,
            destructive_hint = false,
            idempotent_hint = true,
            open_world_hint = true
        )
    )]
    async fn mouse_scroll(
        &self,
        Parameters(MouseScroll { dx, dy }): Parameters<MouseScroll>,
    ) -> String {
        self.exec("input.mouse", json!({"action": "scroll", "dx": dx, "dy": dy}),).await
    }

    #[tool(
        name = "click_coordinate",
        description = "Move to pixel coordinates and click.",
        annotations(
            read_only_hint = false,
            destructive_hint = true,
            idempotent_hint = false,
            open_world_hint = true
        )
    )]
    async fn click_coordinate(
        &self,
        Parameters(ClickCoord { x, y, button }): Parameters<ClickCoord>,
    ) -> String {
        self.call(do_click_coordinate(&self.state, x, y, &button)).await
    }

    #[tool(
        name = "drag",
        description = "Click-and-drag between two pixel coordinates.",
        annotations(
            read_only_hint = false,
            destructive_hint = true,
            idempotent_hint = false,
            open_world_hint = true
        )
    )]
    async fn drag(
        &self,
        Parameters(Drag {
            from_x,
            from_y,
            to_x,
            to_y,
            button,
        }): Parameters<Drag>,
    ) -> String {
        self.call(do_drag(&self.state, from_x, from_y, to_x, to_y, &button)).await
    }
    };
}
